exercises/practice/zebra-puzzle/.meta/example.c

```diff
@@ -0,0 +1,203 @@
+                  continue;
+
+               // 9. Milk is drunk in the middle house.
+               if (house_by_beverage[MILK] != NUM_HOUSES / 2)
+                  continue;
+
+               int brands[] = { CHESTERFIELD, KOOL, LUCKY_STRIKE, OLD_GOLD,
+                                PARLIAMENT };
+               int house_by_brand[] = { 0, 1, 2, 3, 4 };
+               do {
+                  // 7. The Old Gold smoker owns snails.
+                  if (house_by_brand[OLD_GOLD] != house_by_pet[SNAILS])
+                     continue;
+
+                  // 8. Kools are smoked in the yellow house.
+                  if (house_by_brand[KOOL] != house_by_color[YELLOW])
+                     continue;
+
+                  // 11. The man who smokes Chesterfields lives in the
+                  //     house next to the man with the fox.
+                  if (abs(house_by_brand[CHESTERFIELD] - house_by_pet[FOX]) !=
+                      1)
+                     continue;
+
+                  // 12. Kools are smoked in the house next to the house
+                  //     where the horse is kept.
+                  if (abs(house_by_brand[KOOL] - house_by_pet[HORSE]) != 1)
+                     continue;
+
+                  // 13. The Lucky Strike smoker drinks orange juice.
+                  if (house_by_brand[LUCKY_STRIKE] !=
+                      house_by_beverage[ORANGE_JUICE])
+                     continue;
+
+                  // 14. The Japanese smokes Parliaments.
+                  if (house_by_nationality[JAPANESE] !=
+                      house_by_brand[PARLIAMENT])
+                     continue;
+
+                  assert(num_solutions == 0);
+                  ++num_solutions;
+                  drinks_water = nationalities[house_by_beverage[WATER]];
+                  owns_zebra = nationalities[house_by_pet[ZEBRA]];
+
+               } while (next_permutation(brands, house_by_brand));
+            } while (next_permutation(beverages, house_by_beverage));
+         } while (next_permutation(pets, house_by_pet));
+      } while (next_permutation(nationalities, house_by_nationality));
+   } while (next_permutation(colors, house_by_color));
+
+   assert(num_solutions == 1);
+   switch (drinks_water) {
+      case ENGLISH:
+         *out_drinks_water = "English";
+         break;
+      case JAPANESE:
+         *out_drinks_water = "Japanese";
+         break;
+      case NORWEGIAN:
+         *out_drinks_water = "Norwegian";
+         break;
+      case SPANISH:
+         *out_drinks_water = "Spanish";
+         break;
+      case UKRAINIAN:
+         *out_drinks_water = "Ukrainian";
+         break;
+      default:
+         assert(false);
+   }
+   switch (owns_zebra) {
+      case ENGLISH:
+         *out_owns_zebra = "English";
+         break;
+      case JAPANESE:
+         *out_owns_zebra = "Japanese";
+         break;
+      case NORWEGIAN:
+         *out_owns_zebra = "Norwegian";
+         break;
```

Comment:
Why solve the same problem twice?
Is it possible to solve it once and then just query the results (possibly in some generic way) ?

Reply:
Do you mean to take an object oriented approach to the exercise where solving and querying the solution are separate steps?

Reply:
I agree, and don't like it either.  
I was under the impression that these two separate functions would be a what the author of the [problem-specification](https://github.com/exercism/problem-specifications/blob/main/exercises/zebra-puzzle/canonical-data.json) intended, with the two "properties" `drinksWater` and `ownsZebra`.  
I looked at the [C# version](https://github.com/exercism/csharp/blob/main/exercises/practice/zebra-puzzle/ZebraPuzzle.cs) where those are two `static` methods and the [Python version](https://github.com/exercism/python/blob/main/exercises/practice/zebra-puzzle/zebra_puzzle.py) where those are two free functions.

But now that you asked I dug a little further and found the [Go version](https://github.com/exercism/go/blob/main/exercises/practice/zebra-puzzle/zebra_puzzle.go) that has a single function `SolvePuzzle()` which I like more.

I could change the code to something like
```c
struct solution
{
   const char *drinks_water;
   const char *owns_zebra;
};

struct solution solve_puzzle(void);
```